modules/percolator/src/main/java/org/elasticsearch/percolator/PercolatorFieldMapper.java

```diff
@@ -194,35 +236,126 @@ Query percolateQuery(PercolateQuery.QueryStore queryStore, BytesReference docume
             return new PercolateQuery(queryStore, documentSource, candidateMatchesQuery, searcher, verifiedMatchesQuery);
         }
 
-        Query createCandidateQuery(IndexReader indexReader) throws IOException {
+        Query createCandidateQuery(IndexReader indexReader, DocumentMapper docMapper) throws IOException {
+            List<InetAddress> inetValues = new ArrayList<>();
+            List<Long> longValues = new ArrayList<>();
+            List<Double> doubleValues = new ArrayList<>();
+
             List<BytesRef> extractedTerms = new ArrayList<>();
             LeafReader reader = indexReader.leaves().get(0).reader();
             for (FieldInfo info : reader.getFieldInfos()) {
                 Terms terms = reader.terms(info.name);
-                if (terms == null) {
-                    continue;
+                if (terms != null) {
+                    BytesRef fieldBr = new BytesRef(info.name);
+                    TermsEnum tenum = terms.iterator();
+                    for (BytesRef term = tenum.next(); term != null; term = tenum.next()) {
+                        BytesRefBuilder builder = new BytesRefBuilder();
+                        builder.append(fieldBr);
+                        builder.append(FIELD_VALUE_SEPARATOR);
+                        builder.append(term);
+                        extractedTerms.add(builder.toBytesRef());
+                    }
+                } else if (info.getPointDimensionCount() != 0) {
+                    FieldMapper fieldMapper = docMapper.mappers().getMapper(info.name);
+                    if (fieldMapper == null) {
+                        continue;
+                    }
+
+                    String fieldTypeName;
+                    if (fieldMapper instanceof NumberFieldMapper) {
+                        fieldTypeName = ((NumberFieldMapper) fieldMapper).fieldType().typeName();
+                    } else if (fieldMapper instanceof DateFieldMapper) {
+                        fieldTypeName = "long";
+                    } else if (fieldMapper instanceof IpFieldMapper) {
+                        fieldTypeName = "ip";
+                    } else if (fieldMapper instanceof SeqNoFieldMapper) {
+                        // This is an internal field, which we should ignore.
+                        continue;
+                    } else {
+                        throw new IllegalStateException("Unexpected fieldMapper [" + fieldMapper + "]");
+                    }
```

Comment:
I don't think we need to distinguish? We don't need to know what the bytes represent, we only rely on the fact that they maintain order?